### PR TITLE
Issue215:  Export NUnit XML Test results

### PR DESCRIPTION
**Pull request by OsirisTerje**
PR for Issue #215

### CODE REVIEW

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 1]
```diff
@@ -301,6 +315,20 @@ public void Load(string settingsXml)
 
             // Update NumberOfTestWorkers based on the DisableParallelization and NumberOfTestWorkers from runsettings.
             UpdateNumberOfTestWorkers();
+
+
+            string ValidatedPath(string path, string what)
+            {
+                try
+                {
+                    return Path.GetFullPath(path);
```
[the remark this answers is not in the document]
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: Yes, the engine comes along with the package.   
About the path:  If I add a path to a non-existing folder, the engine crashes.  So I have to ensure it does exist, and then I think it is better to resolve it up front.  Agree with the WorkDirectory setting.

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 2]
```diff
@@ -62,6 +62,8 @@ public interface IAdapterSettings
         string DefaultTestNamePattern { get; }
 
         VsTestCategoryType VsTestCategoryType { get; }
+        string TestOutput { get; }
+        bool UseTestOutput { get; }
```
[the remark this answers is not in the document]
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: Ok, I'll make them TestOutputXml, and UseTestOutputXml

File: src/NUnitTestAdapter/AdapterSettings.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CharliePoole) [reply src/NUnitTestAdapter/AdapterSettings.cs]: That should be a bug to the engine as well.

I know that the result-writer "extension" is built into the engine, but is it actually initialized? I didn't think it was but if this code is working, I suppose it must be.

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 3]
```diff
@@ -221,14 +225,18 @@ public void Load(string settingsXml)
...
             WorkDirectory = GetInnerTextWithLog(nunitNode, nameof(WorkDirectory));
             DefaultTimeout = GetInnerTextAsInt(nunitNode, nameof(DefaultTimeout), 0);
             NumberOfTestWorkers = GetInnerTextAsInt(nunitNode, nameof(NumberOfTestWorkers), -1);
             ShadowCopyFiles = GetInnerTextAsBool(nunitNode, nameof(ShadowCopyFiles), false);
             UseVsKeepEngineRunning = GetInnerTextAsBool(nunitNode, nameof(UseVsKeepEngineRunning), false);
             BasePath = GetInnerTextWithLog(nunitNode, nameof(BasePath));
             PrivateBinPath = GetInnerTextWithLog(nunitNode, nameof(PrivateBinPath));
+            var testOutput = GetInnerTextWithLog(nunitNode, nameof(TestOutputXml));
+            if (!string.IsNullOrEmpty(testOutput))
+                TestOutputXml = ValidatedPath(testOutput,nameof(TestOutputXml));
```
Comment (jnm2) [inline src/NUnitTestAdapter/AdapterSettings.cs]: Typically there's a space after the comma.

Comment (jnm2) [inline src/NUnitTestAdapter/AdapterSettings.cs]: Typically there's a space after the comma.

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 4]
```diff
@@ -248,14 +256,15 @@ public void Load(string settingsXml)
...
+                        _logger.Warning(
+                            $"Invalid value ({vsTestCategoryType}) for VsTestCategoryType, should be either NUnit or MsTest");
                         break;
                 }
 
 
 
 #if SUPPORT_REGISTRY_SETTINGS
-            // Legacy (CTP) registry settings override defaults
+// Legacy (CTP) registry settings override defaults
```
Comment (jnm2) [inline src/NUnitTestAdapter/AdapterSettings.cs]: Why these changes?

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 5]
```diff
@@ -266,17 +275,21 @@ public void Load(string settingsXml)
...
+// Force Verbosity to 1 under Debug
             Verbosity = 1;
 #endif
 
-            var inProcDataCollectorNode = doc.SelectSingleNode("RunSettings/InProcDataCollectionRunSettings/InProcDataCollectors");
-            InProcDataCollectorsAvailable = inProcDataCollectorNode != null && inProcDataCollectorNode.SelectNodes("InProcDataCollector").Count > 0;
+            var inProcDataCollectorNode =
+                doc.SelectSingleNode("RunSettings/InProcDataCollectionRunSettings/InProcDataCollectors");
+            InProcDataCollectorsAvailable = inProcDataCollectorNode != null &&
+                                            inProcDataCollectorNode.SelectNodes("InProcDataCollector").Count > 0;
```
Comment (jnm2) [inline src/NUnitTestAdapter/AdapterSettings.cs]: (Seems like a lot of changes are just rewrapping? I usually do formatting in a separate commit from the logical changes.)

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 6]
```diff
@@ -301,6 +315,20 @@ public void Load(string settingsXml)
...
             UpdateNumberOfTestWorkers();
+
+
+            string ValidatedPath(string path, string what)
+            {
+                try
+                {
+                    return Path.GetFullPath(path);
+                }
+                catch (Exception)
```
Comment (jnm2) [inline src/NUnitTestAdapter/AdapterSettings.cs]: `(Exception)` is redundant since the runtime wraps anything thrown into an Exception object.

File: src/NUnitTestAdapter/AdapterSettings.cs [hunk 7]
```diff
@@ -301,6 +315,20 @@ public void Load(string settingsXml)
 
             // Update NumberOfTestWorkers based on the DisableParallelization and NumberOfTestWorkers from runsettings.
             UpdateNumberOfTestWorkers();
+
+
+            string ValidatedPath(string path, string what)
```
Comment (jnm2) [inline src/NUnitTestAdapter/AdapterSettings.cs]: `what` could be `purpose`?

File: src/NUnitTestAdapter/NUnit3TestExecutor.cs [hunk 8]
```diff
@@ -329,6 +330,23 @@ private void RunAssembly(string assemblyPath, TestFilter filter)
             }
         }
 
+
+        private void GenerateTestOutput(XmlNode testResults, string assemblypath)
+        {
+            if (!Settings.UseTestOutputXml)
+                return;
+#if NETCOREAPP1_0
+#else
```
Comment (jnm2) [inline src/NUnitTestAdapter/NUnit3TestExecutor.cs]: `#if !NETCOREAPP1_0`?

File: src/NUnitTestAdapter/NUnit3TestExecutor.cs [hunk 9]
```diff
@@ -329,6 +330,23 @@ private void RunAssembly(string assemblyPath, TestFilter filter)
...
+            var path = Path.Combine(Settings.TestOutputXml,$"{Path.GetFileNameWithoutExtension(assemblypath)}.xml");
```
Comment (jnm2) [inline src/NUnitTestAdapter/NUnit3TestExecutor.cs]: Space usually after comma.
Also, `Path.ChangeExtension` is an alternative to string templating.

File: src/NUnitTestAdapter/NUnit3TestExecutor.cs [hunk 10]
```diff
@@ -329,6 +330,23 @@ private void RunAssembly(string assemblyPath, TestFilter filter)
             }
         }
 
+
+        private void GenerateTestOutput(XmlNode testResults, string assemblypath)
```
Comment (jnm2) [inline src/NUnitTestAdapter/NUnit3TestExecutor.cs]: Should be camel-cased: `assemblyPath`

File: src/NUnitTestAdapter/NUnitTestAdapter.cs [hunk 11]
```diff
@@ -138,6 +138,16 @@ protected void Initialize(IDiscoveryContext context, IMessageLogger messageLogge
             }
         }
 
+        private void CheckDirectories()
+        {
+            bool ok = true;
```
Comment (jnm2) [inline src/NUnitTestAdapter/NUnitTestAdapter.cs]: Unused local

File: src/NUnitTestAdapter/NUnitTestAdapter.cs [hunk 12]
```diff
@@ -125,11 +125,11 @@ protected void Initialize(IDiscoveryContext context, IMessageLogger messageLogge
             TestLog = new TestLogger(messageLogger);
             Settings = new AdapterSettings(TestLog);
             TestLog.InitSettings(Settings);
-
```
Comment (jnm2) [inline src/NUnitTestAdapter/NUnitTestAdapter.cs]: You don't prefer this blank line?

File: src/NUnitTestAdapter/AdapterSettings.cs (hunk 6 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: Not according to R#

File: src/NUnitTestAdapter/NUnit3TestExecutor.cs (hunk 8 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/NUnit3TestExecutor.cs]: No, the else block It should only be run if .net framework, not .net core.  Missing a #ifndef directive

File: src/NUnitTestAdapter/NUnitTestAdapter.cs (hunk 12 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/NUnitTestAdapter.cs]: No :-)

File: src/NUnitTestAdapter/AdapterSettings.cs (hunk 4 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: Fixed

File: src/NUnitTestAdapter/AdapterSettings.cs (hunk 3 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: Fixed

File: src/NUnitTestAdapter/AdapterSettings.cs (hunk 7 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: Fixed

File: src/NUnitTestAdapter/NUnit3TestExecutor.cs (hunk 9 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/NUnit3TestExecutor.cs]: Yes, that's better

File: src/NUnitTestAdapter/NUnitTestAdapter.cs (hunk 11 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/NUnitTestAdapter.cs]: Fixed

File: src/NUnitTestAdapter/NUnit3TestExecutor.cs (hunk 8 above)
Comment (jnm2) [reply src/NUnitTestAdapter/NUnit3TestExecutor.cs]: @OsirisTerje `#if !NETCOREAPP1_0` is that directive. I'm confused because I don't see a difference between `#if !NETCOREAPP1_0` and what you have, except that it's shorter.

Comment (OsirisTerje) [reply src/NUnitTestAdapter/NUnit3TestExecutor.cs]: You're right.  Will fix.

File: src/NUnitTestAdapter/AdapterSettings.cs (hunk 6 above)
Comment (OsirisTerje) [reply src/NUnitTestAdapter/AdapterSettings.cs]: You're right here too of course, but then I should change the dotsettings file.  I'll have a look at do that if I don't see any side effects.
